### DISCUSSION TITLE
is it possible to sort column name using SQL

### DISCUSSION THREAD

**Discussion by djouallah** [duckdb/duckdb#12047] [General]
I am using this and it works fine, but I need a pure SQL solution, is it possible ?

```
final=duckdb.sql(""" SELECT * from table1   """)
columns_sorted = sorted(final.columns)
df=final.select(*columns_sorted)
```

**Comment by soerenwolfers** [comment]
See "Using DESCRIBE in a subquery" https://duckdb.org/docs/guides/meta/describe.html

**Reply by cmdlineluser** [reply]
Very interesting!

Is it actually possible to select a "dynamically generated" list of column names?

```python
duckdb.sql("""
with 
   subquery as (
      from (values ('d1', 'y1', 'm1')) t(date, year, month)
   ),
   sorted_column_names as (
      from (describe from subquery)
      select column_name
      order by all
   )
from subquery
-- select sorted_column_names <- not possible to pass dynamic column names?
""")

# ┌─────────┬─────────┬─────────┐
# │  date   │  year   │  month  │
# │ varchar │ varchar │ varchar │
# ├─────────┼─────────┼─────────┤
# │ d1      │ y1      │ m1      │
# └─────────┴─────────┴─────────┘
```

**Reply by soerenwolfers** [reply]
Yes, it's possible using json_serialize_sql and json_execute_serialized_sql. See https://duckdb.org/docs/extensions/json.html

**Reply by soerenwolfers** [reply]
I am almost sure I also recently saw a new feature that allows dynamic SQL execution more directly, but can't find a link.

**Reply by cmdlineluser** [reply]
@soerenwolfers There was the recentish blog post that had the `MACRO` stuff in it.

https://duckdb.org/2024/03/01/sql-gymnastics.html

Which did have something similar:

> Selecting a dynamic list of columns by combining the `COLUMNS` expression with a lambda and the `list_contains` function

But as the lambda is just accept/reject, it didn't seem like something that could help with specific ordering.

**Reply by Tishj** [reply]
Probably `query` which landed on `feature`

**Reply by soerenwolfers** [reply]
Ah yes, thanks https://github.com/duckdb/duckdb/pull/10586

**Reply by cmdlineluser** [reply]
Ah, fascinating.

So `feature` is a separate branch - that would explain why I couldn't find those functions. ^_^
